lib/pbench/server/database/models/datasets.py

```diff
@@ -253,7 +254,7 @@ class Dataset(Database.Base):
 
     Columns:
         id          Generated unique ID of table row
-        owner_id    Owning UUID of the owner of the dataset
+        owner       Owning username of the dataset
```

Comment:
Yeah I wonder about that, theoretically, we should provide the delete user functionality through our API, if the user wishes to remove their existence on the Pbench-server then we should honor the request, isn't it? When the user is deleted all the datasets owned by the user become public. I should make the foreign key user_id `nullable=True` _I think_.

Reply:
Deleting a cached `User` shouldn't have anything to do with the lifetime (or ownership) of a dataset; and that's a problem with re-instating the relationship, although it's convenient in other ways.

But maybe we just don't allow deleting your cached user identity until you've changed the ownership of all datasets you own? Eventually, we might allow changing ownership to a *group* rather than an individual; but you could already have all your datasets re-assigned to me if you want to delete all trace of yourself.

Reply:
I think that datasets should be owned by a UUID.  Certainly, they should _not_ be owned by a **username**, as I don't think there is any reason to expect those to be unique.

If we feel we _have_ to, datasets could be owned by a `User` (i.e., associated with a row in that table keyed by UUID), but I'm not convinced that that is a win (although I would certainly bow to you guys with better database knowledge than I possess -- e.g., it would be handy if `User`-ownership gave us automatic translation in terms of providing to the client the owner as a username).

But, as Dave says, if a dataset depends on a `User` entry, then we cannot delete that `User` entry...which is probably fine...but, yeah, it means that there's no point in providing a "delete user" operation...unless we _want_ the Big Hammer of being able to delete all datasets along with a `User` in one swell foop (although, we would have to handle the ES data, too 😛).

> theoretically, we should provide the delete user functionality through our API, if the user wishes to remove their existence on the Pbench-server then we should honor the request, isn't it?

A better approach would be to keep nothing worth deleting about a user on the Pbench Server.  To my way of thinking, the only "stuff" that a user should have on the Server, other than datasets, is their UUID-to-username translation and any API access keys which they've requested.  There might be "settings" (like default retention period), but there shouldn't be any PII -- that all comes from somewhere else.

> When the user is deleted all the datasets owned by the user become public.

A better approach would be to refuse to delete a user which owns any datasets -- instead, require that the datasets be shifted to new owners or deleted, first.  (Force-publishing datasets is worse than refusing to delete the user....)

Reply:
> If we feel we _have_ to, datasets could be owned by a `User` (i.e., associated with a row in that table keyed by UUID), but I'm not convinced that that is a win

Using `User` is only a minor win in that it gives us a straightforward SQLAlchemy foreign key and relationship so that a returned `Dataset` has the `User` (and therefore both OIDC ID and username) immediately available.

It's worth noting that in SQL-speak, the `ForeignKey("users.id")` is entirely arbitrary. We just need a unique column in `User` and we could just as easily use `owner_id = Column(String, ForeignKey("users.oidc_id"))`; though, since SQLAlchemy will implicitly do the `join` due to the `relationship`, the distinction isn't very important.

Reply:
Given that we have the UUID column and that we depend upon it being unique, I don't see any point to having a `"users.id"` column, and having the relationship based on the ODIC id seems good...because it implicitly gives us the `username` for a dataset...sounds great!

Reply:
We still have the "typo" here in the docstring ... 😁